gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+    """
+    Lumerical EME simulation
+
+    Attributes:
+        component: Component geometry to simulate
+        layerstack: PDK layerstack
+        session: Lumerical session
+        simulation_settings: EME simulation settings
+        convergence_settings: EME convergence settings
+        dirpath: Directory where simulation files are saved
+        mesh_convergence_data: Mesh convergence results
+        cell_convergence_data: Cell convergence results
+        mode_convergence_data: Mode convergence results
+
+    """
+
+    def __init__(
+        self,
+        component: Component,
+        layerstack: LayerStack | None = None,
+        session: lumapi.MODE | None = None,
+        simulation_settings: SimulationSettingsLumericalEme = LUMERICAL_EME_SIMULATION_SETTINGS,
+        convergence_settings: ConvergenceSettingsLumericalEme = LUMERICAL_EME_CONVERGENCE_SETTINGS,
+        dirpath: PathType | None = "",
+        hide: bool = True,
+        run_mesh_convergence: bool = False,
+        run_cell_convergence: bool = False,
+        run_mode_convergence: bool = False,
+        run_overall_convergence: bool = False,
+        **settings,
+    ):
+        """
+        Set up EME simulation based on component geometry and simulation settings. Optionally, run convergence.
+
+        Parameters:
+            component: Component geometry to simulate
+            layerstack: PDK layerstack
+            session: Lumerical session
+            simulation_settings: EME simulation settings
+            convergence_settings: EME convergence settings
+            dirpath: Directory where simulation files are saved
+            hide: Hide simulation if True, else show GUI
+            run_mesh_convergence: If True, run sweep of mesh and monitor sparam convergence.
+            run_cell_convergence: If True, run sweep of number of cells in central group span and monitor sparam convergence.
+            run_mode_convergence: If True, run sweep of number of modes and monitor sparam convergence.
+            run_overall_convergence: If True, run combination of mesh and cell convergence to quicken convergence.
+        """
+        # Set up variables
+        dirpath = dirpath or Path(__file__).resolve().parent
+        simulation_settings = dict(simulation_settings)
+
+        if hasattr(component.info, "simulation_settings"):
+            simulation_settings |= component.info.simulation_settings
+            logger.info(
+                f"Updating {component.name!r} sim settings {component.simulation_settings}"
+            )
+        for setting in settings:
+            if setting not in simulation_settings:
+                raise ValueError(
+                    f"Invalid setting {setting!r} not in ({list(simulation_settings.keys())})"
+                )
+
+        simulation_settings.update(**settings)
+        ss = SimulationSettingsLumericalEme(**simulation_settings)
+
+        # Check number of cell groups are aligned
```

Comment:
this should probably go after session initialization? broadly, this is a very long function that maybe could be broken up a little?

Reply:
drawing component geometry should definitely be its own method